src/DeviceInterface.cpp

```diff
@@ -0,0 +1,133 @@
+#include "JITModule.h"
+#include "StmtCompiler.h"
+#include "Target.h"
+
+#include "runtime/HalideRuntime.h"
+#include "runtime/HalideRuntimeCuda.h"
+#include "runtime/HalideRuntimeOpenCL.h"
+#include "runtime/HalideRuntimeOpenGL.h"
+
+using namespace Halide;
+using namespace Halide::Internal;
+
+namespace {
+
+template <typename fn_type>
+bool lookup_runtime_routine(const char *name, const Target &target, fn_type &result) {
+    StmtCompiler stmt_compiler(target);
+    std::vector<JITModule> runtime(JITSharedRuntime::get(stmt_compiler.get_codegen(), target));
```

Comment:
There's an issue here that isn't an immediate problem. I found the issue because in my modules branch, StmtCompiler is gone and JITSharedRuntime doesn't depend on a CodeGen_LLVM object. 

The issue is that this will attempt to create a new runtime, even if the only reason lookup_runtime_routine is being called is to call halide_device_free on a buffer_t that doesn't have a dev object.

My solution in that branch was to add a parameter to JITSharedRuntime 'bool create'. 'create' is only true when the function being requested might be able to do useful work even if the GPU runtime has not already been loaded, e.g. halide_device_malloc. The reasoning being that the only reason halide_device_free, halide_device_release, etc. could possibly have any work to do is if the GPU runtime has been loaded already (otherwise halide_device_malloc, etc. could never have been called).

Again, this is not an immediate problem, it doesn't break anything in this branch. It's only a potential performance issue. Since I have a fix incoming in hopefully the next PR, my suggestion would be to leave it be for now.